requirements.txt

```diff
@@ -1,6 +1,6 @@
 joblib
-numpy
+numpy<1.20.0
 pandas
-scikit-learn>=0.22.0
+scikit-learn>=0.22.0,<0.24.0
```

Comment:
remind me again what's wrong with v24? 

sklearn 24 has lots of goodies, so I wonder if it will be possible to make a complete jump to that newer version in mab, and across libs.

Reply:
@filip-michalsky why are you adding the <0.24 requirement? The change in StandardScaler is an issue for use cases, not for the library

Reply:
I was not able to successfully run unit tests until sklearn was bumped down below 0.24. Replicated issue on both on my machine and a C2C box

Reply:
namely, with sklearn 0.24.0+, 6 tests fail:
======================================================================
FAIL: test_fit_twice_thompson_thresholds (test_clusters.ClustersTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/shared/corpid/a649154/pr110447-ai_coe_pi_atlas/projects/sr/PRTCJP-756/libraries/mabwiser/tests/test_clusters.py", line 274, in test_fit_twice_thompson_thresholds
    self.assertListEqual(arms, [3, 4])
AssertionError: Lists differ: [3, 3] != [3, 4]

First differing element 1:
3
4

- [3, 3]
?     ^

+ [3, 4]
?     ^


======================================================================
FAIL: test_greedy0_n3 (test_clusters.ClustersTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/shared/corpid/a649154/pr110447-ai_coe_pi_atlas/projects/sr/PRTCJP-756/libraries/mabwiser/tests/test_clusters.py", line 105, in test_greedy0_n3
    self.assertListEqual(arms, [2, 2])
AssertionError: Lists differ: [2, 3] != [2, 2]

First differing element 1:
3
2

- [2, 3]
?     ^

+ [2, 2]
?     ^


======================================================================
FAIL: test_partial_fit_thompson_thresholds (test_clusters.ClustersTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/shared/corpid/a649154/pr110447-ai_coe_pi_atlas/projects/sr/PRTCJP-756/libraries/mabwiser/tests/test_clusters.py", line 234, in test_partial_fit_thompson_thresholds
    self.assertListEqual(arms, [3, 4])
AssertionError: Lists differ: [3, 3] != [3, 4]

First differing element 1:
3
4

- [3, 3]
?     ^

+ [3, 4]
?     ^


======================================================================
FAIL: test_UCB1_c2 (test_parallel.ParallelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/shared/corpid/a649154/pr110447-ai_coe_pi_atlas/projects/sr/PRTCJP-756/libraries/mabwiser/tests/test_parallel.py", line 239, in test_UCB1_c2
    self.assertEqual(arm,  [[3, 3, 3, 1, 1, 1, 3, 1, 3, 3] for _ in range(5)])
AssertionError: Lists differ: [[3, 3, 1, 1, 1, 1, 3, 1, 3, 3], [3, 3, 1, 1, 1, [106 chars], 3]] != [[3, 3, 3, 1, 1, 1, 3, 1, 3, 3], [3, 3, 3, 1, 1, [106 chars], 3]]

First differing element 0:
[3, 3, 1, 1, 1, 1, 3, 1, 3, 3]
[3, 3, 3, 1, 1, 1, 3, 1, 3, 3]

Diff is 645 characters long. Set self.maxDiff to None to see it.

======================================================================
FAIL: test_UCB1_c2_backend (test_parallel.ParallelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/shared/corpid/a649154/pr110447-ai_coe_pi_atlas/projects/sr/PRTCJP-756/libraries/mabwiser/tests/test_parallel.py", line 882, in test_UCB1_c2_backend
    self.assertEqual(arm, [[3, 3, 3, 1, 1, 1, 3, 1, 3, 3] for _ in range(5)])
AssertionError: Lists differ: [[3, 3, 1, 1, 1, 1, 3, 1, 3, 3], [3, 3, 1, 1, 1, [106 chars], 3]] != [[3, 3, 3, 1, 1, 1, 3, 1, 3, 3], [3, 3, 3, 1, 1, [106 chars], 3]]

First differing element 0:
[3, 3, 1, 1, 1, 1, 3, 1, 3, 3]
[3, 3, 3, 1, 1, 1, 3, 1, 3, 3]

Diff is 645 characters long. Set self.maxDiff to None to see it.

======================================================================
FAIL: test_thompson_n3 (test_parallel.ParallelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/shared/corpid/a649154/pr110447-ai_coe_pi_atlas/projects/sr/PRTCJP-756/libraries/mabwiser/tests/test_parallel.py", line 579, in test_thompson_n3
    self.assertListEqual(arms, [1, 1, 3, 1, 1, 1, 1, 2, 1, 2])
AssertionError: Lists differ: [2, 1, 3, 3, 3, 2, 2, 3, 2, 3] != [1, 1, 3, 1, 1, 1, 1, 2, 1, 2]

First differing element 0:
2
1

- [2, 1, 3, 3, 3, 2, 2, 3, 2, 3]
+ [1, 1, 3, 1, 1, 1, 1, 2, 1, 2]

----------------------------------------------------------------------
Ran 472 tests in 184.675s

FAILED (failures=6)

Reply:
I also confirmed that the 6 tests fail when scikit-learn jumps to 0.24+. They all seem related to a change in behavior in KMeans. I could not find any hints in the scikit-learn [changelog](https://scikit-learn.org/stable/whats_new/v0.24.html#version-0-24-0) that would explain the change though. Since it does not break any code and is a reproducibility issue I don't think it's necessary to explicitly restrict the version, but will need to figure out what we do with the tests.

Reply:
Ok I was able to reproduce the issue, looks like something changed with KMeans clustering again

Reply:
yeah, I thought restricting the version of `sklearn` is a quick fix for this and the only alternative is to figuring out what changed in sklearn under the hood ...

Reply:
Looks like they did a complete rewrite of KMeans: https://github.com/scikit-learn/scikit-learn/commit/769794249c2e361c15ae15f4de8adcc02976ae74

Reply:
Well that would explain it.

Reply:
again, disclaimer: I am not close to this work so my comments might be off..

I am not comfortable with the < 2.4 requirements. If they rewrite the KMeans, well that breaks our code.. which is natural, and here the tests are failing.. I would rather we update the tests and make the jump to newer versions of the library, no? 

We don't have "corporate customers" to maintain behavior for. Let's use that luxury and uptake new stuff?

Reply:
Agreed @skadio ! The reason I proposed the version restrictions is to propose a quick solution - albeit updating things on our end with a version bump makes more sense! I will send an update to this PR.